src/coreclr/jit/emitarm64.cpp

```diff
@@ -15629,24 +15638,58 @@ bool emitter::IsRedundantMov(instruction ins, emitAttr size, regNumber dst, regN
         return false;
     }
 
+    const bool isFirstInstrInBlock = (emitCurIGinsCnt == 0) && ((emitCurIG->igFlags & IGF_EXTEND) == 0);
+
     if (dst == src)
     {
         // A mov with a EA_4BYTE has the side-effect of clearing the upper bits
         // So only eliminate mov instructions that are not clearing the upper bits
         //
-        if (isGeneralRegisterOrSP(dst) && (size == EA_8BYTE))
+        if (isGeneralRegisterOrSP(dst) && (size == EA_8BYTE) && !isSignExtendMov)
         {
             JITDUMP("\n -- suppressing mov because src and dst is same 8-byte register.\n");
             return true;
         }
-        else if (isVectorRegister(dst) && (size == EA_16BYTE))
+        else if (isVectorRegister(dst) && (size == EA_16BYTE) && !isSignExtendMov)
         {
             JITDUMP("\n -- suppressing mov because src and dst is same 16-byte register.\n");
             return true;
         }
+        else if ((emitLastIns != nullptr) && !isFirstInstrInBlock)
+        {
+            const instruction prevIns = emitLastIns->idIns();
+            // See if the previous instruction already cleared upper 4 bytes for us unintentionally
+            if ((emitLastIns->idReg1() == dst) && (emitLastIns->idOpSize() == EA_4BYTE))
+            {
+                if (!isSignExtendMov && (size == EA_4BYTE) &&
+                    ((prevIns == INS_ldr) || (prevIns == INS_ldrh) || (prevIns == INS_ldrb)))
+                {
+                    JITDUMP("\n -- suppressing mov because ldr already cleared upper 4 bytes\n");
+                    return true;
+                }
+                else if (isSignExtendMov && (size == EA_8BYTE) && ((prevIns == INS_ldr) || (prevIns == INS_ldrsw) ||
+                                                                   (prevIns == INS_ldrsh) || (prevIns == INS_ldrsb)))
+                {
+                    // Special case: upgrade ldr to ldrsw
+                    if (prevIns == INS_ldr)
+                    {
+                        emitLastIns->idIns(INS_ldrsw);
+                        JITDUMP("\n -- suppressing sxtw and chaning previous ldr to ldrsw\n");
+                    }
+                    else
+                    {
+                        JITDUMP("\n -- suppressing sxtw because ldrs already cleared upper 4 bytes\n");
+                    }
+                    return true;
+                }
+            }
+        }
```

Comment:
it's done in the emitter just to estimate impact and check tests so no review is required at this point